enhancements/multi-arch/heterogeneous-architecture.md

```diff
@@ -0,0 +1,383 @@
+OLM would need to make sure that it deploys operators only to nodes whose architecture matches the architectures on which the operator is supported.
+Today, OLM has a subscription object which is used to tell it how to deploy an operator. It has support for the nodeselector field but this would
+not be sufficient to specify filtering on multiple architectures. To this end, it would have to support the podAffinity field which uses the
+nodeselectorterms to filter on multiple values.
+
+Instead of having OLM dictate the operator deployment on specific nodes, the individual operators themselves would specify which architectures
+they support through a nodeselector object which OLM will then provide the ability for the user to override if necessary. This information should
+also be added to the operands and potentially be made configurable by the user requesting the operand. The operator-sdk framework could also be
+potentially leveraged to support the generation of this supported architecture list when creating operators.
+
+### Machine Config Operator
+
+No changes are needed in the MCO for Phase 1 as the machine-os-content image would be a manifest listed image and the machine config daemon would
+extract the relevant architecture's machine-os-content based on the node it runs on. In the future, architecture specific configurations should
+not be required, instead they need to be templatized and be generalized for certain items, for example kubelet system reserved memory could scale
+based on page size.
+
+### Machine API
+
+- As of today, machinesets can be mutated to change the instance type and image id and this could potentially result in inconsistency between existing
+machinesets and newly created machinesets. If the instance type and image is changed to one of a different architecture, there would be workers of
+different architectures with the same machineset definition. While this may not be a problem, it doesn't make for a good user experience. Hence,
+machineset specs will not allowed to be mutated. Anytime a change is required, a new machineset needs to be created and the old one needs to be scaled
+down.
+- Changes are needed in the autoscaler to make it aware of the node architecture when autoscaling from zero. The `kubernetes-io-arch` label needs to be
+set when building the NodeInfo object so the autoscaler can pick the correct pool to scale up for allocating nodes of a particular architecture.
+
+### Hypershift
+
+The addition of heterogeneous support in Openshift also opens up a new dimension for [support][OCPPLAN-5684] in Hypershift where the control plane 
+and the workers could run on different architectures. There would be certain changes needed to realize this as there are limitations which need to
+be addressed:
+
+- When creating the hosted cluster, the latest x86 release payload is picked by default. This would have to change to pick a heterogeneous payload
+when creating the hosted cluster on mixed architecture workers.
```

Comment:
Why wouldn't the selection be based on the channel, as described above in the CVO section?

Reply:
are you suggesting that hypershift should read the pullsepc from the clusterversion object ? today the hosted control plane image defaults to picking an x86 image: https://github.com/openshift/hypershift/blob/main/cmd/version/version.go#L17

Reply:
to be clear, the hosted cluster creation uses the management cluster's payload -  so if the management cluster is heterogeneous, we don't have a problem. But if the management cluster is of a single arch , we cannot create a nodepool/hosted cluster of a different arch without tweaks

Reply:
> to be clear, the hosted cluster creation uses the management cluster's payload 

no it doesn't.  When creating a guest cluster you pick the payload to use.  Hypershift host/management clusters can install guest/hosted clusters of varying versions, not just the same version as the host/management cluster.

Reply:
that's correct. i meant that by default the management cluster's spec is used. The bigger problem here is - the hosted cluster's ignition server uses the nodepool's mco to generate ignition for the nodepools. Likewise, for some of the components deployed on the nodepool, the hosted cluster's image pullspecs are used. so even if we did have a management cluster and hosted cluster on a single arch, we cannot have a nodepool of a different arch without using a heterogeneous payload.